cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs

```diff
@@ -103,54 +108,81 @@ type FixedPoint = Digits34
 fpPrecision :: FixedPoint
 fpPrecision = (10 :: FixedPoint) ^ (34 :: Integer)
 
+-- | Maximum precision possible for unit interval when backed by a 64bit
+type UnitScale = 19
+
+unitScale :: Int
+unitScale = fromInteger (natVal (Proxy :: Proxy UnitScale))
+
+newtype RawUnit
+  = RawUnit Word64
+  deriving newtype (Show, Eq, Ord, Num, Real, Enum, Integral, NFData)
+
+instance Bounded RawUnit where
+  minBound = 0
+  maxBound = RawUnit (10 ^ unitScale)
+
 -- | Type to represent a value in the unit interval [0; 1]
-newtype UnitInterval = UnsafeUnitInterval (Ratio Word64)
-  deriving (Show, Ord, Eq, Generic)
-  deriving newtype (NoThunks, NFData)
+newtype UnitInterval = UnitInterval
+  { unitDecimal :: Decimal RoundHalfEven UnitScale RawUnit
+  }
+  deriving (Generic)
+  deriving newtype (Show, Ord, Eq, Bounded, NFData)
+
+instance Default UnitInterval where
+  def = minBound
+
+instance NoThunks UnitInterval where
+  noThunks ctx = noThunks ctx . coerce @_ @Word64
+  wNoThunks ctx = wNoThunks ctx . coerce @_ @Word64
 
+-- We are using CBOR serialization through `Ratio Word64` for hisotrical reasons. It would
+-- be better to use Word64 directly: `toCBOR . coerce @_ @Word64`
 instance ToCBOR UnitInterval where
-  toCBOR (UnsafeUnitInterval u) = ratioToCBOR u
+  toCBOR = ratioToCBOR . (% coerce (maxBound :: RawUnit)) . coerce @_ @Word64
 
 instance FromCBOR UnitInterval where
   fromCBOR = do
     r <- ratioFromCBOR
-    case mkUnitInterval r of
+    let n = toInteger (numerator r :: Word64)
+        d = toInteger (denominator r :: Word64)
+    case unitIntervalFromRational (n % d) of
       Nothing -> cborError $ DecoderErrorCustom "UnitInterval" (Text.pack $ show r)
       Just u -> pure u
 
 instance ToJSON UnitInterval where
-  toJSON ui = toJSON (fromRational (unitIntervalToRational ui) :: Scientific)
+  toJSON = toJSON . toScientificDecimal . unitDecimal
 
 instance FromJSON UnitInterval where
   parseJSON v = do
     d <- parseJSON v
-    case mkUnitInterval (realToFrac (d :: Scientific) :: Ratio Word64) of
-      Just u -> return u
-      Nothing -> fail "The value must be between 0 and 1 (inclusive)"
+    either fail pure $ fromScientificUnitInterval d
+
+-- | safe-decimal-0.2.1.0 has a fixed version of `fromScientificDecimalBounded` function
```

Comment:
Is there a reason we can't use it?

Reply:
I would prefer this as well, but I just uploaded this new version on hackage a few days ago and from what I understand, we need to update index-state. I am not sure what's the process on going about it.